fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/properties/FDBRecordContextProperties.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * FDBRecordContextProperties.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2021 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.provider.foundationdb.properties;
+
+import com.apple.foundationdb.record.RecordCoreException;
+import com.apple.foundationdb.record.logging.LogMessageKeys;
+
+import javax.annotation.Nonnull;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+public class FDBRecordContextProperties {
+    private static final FDBRecordContextProperties EMPTY_PROPERTIES = new FDBRecordContextProperties(Collections.emptyMap());
+
+    @Nonnull
+    private final Map<String, FDBPropertyKey<?>> properties;
+
+    private FDBRecordContextProperties(@Nonnull Map<String, FDBPropertyKey<?>> properties) {
+        this.properties = properties;
+    }
+
+    @Nonnull
+    public Map<String, FDBPropertyKey<?>> getProperties() {
+        return properties;
+    }
+
+    @Nonnull
+    public <T> T getProperty(@Nonnull FDBPropertyKey.Builder<T> propBuilder) {
+        if (properties.containsKey(propBuilder.getName())) {
+            Object value = properties.get(propBuilder.getName()).getValue();
+            try {
+                return propBuilder.getType().cast(value);
+            } catch (ClassCastException ex) {
+                throw new RecordCoreException("Invalid type for record context property", ex)
+                        .addLogInfo(LogMessageKeys.PROPERTY_NAME, propBuilder.getName(),
+                                LogMessageKeys.PROPERTY_TYPE, value.getClass().getName());
+            }
+        } else {
+            return propBuilder.getDefaultValue();
+        }
+    }
+
+    public static Builder newBuilder() {
+        return new Builder();
+    }
+
+    public static FDBRecordContextProperties getEmptyInstance() {
+        return EMPTY_PROPERTIES;
+    }
+
+    public static class Builder {
+        private final Map<String, FDBPropertyKey<?>> properties = new HashMap<>();
+
+        public Builder addProp(@Nonnull String name, @Nonnull FDBPropertyKey<?> prop) {
+            properties.put(name, prop);
```

Comment:
Should this throw an exception if the `name` is already in the properties list? The argument for throwing an error is that it might prevent us from accidentally defining two properties with the same name, but the argument against is that overriding the property could be how someone overrides the value of the property

Reply:
Added a hashSet to avoid name duplication.